File: python/equity_call_pricer.py

from fx_generator import FXGenerator
from ir_generator import InterestRateGenerator

import numpy as np

class EquityCallPricer:

    def __init__(
        self,
        strike: float,
        start_val: float,
        vol: float,
        points_per_path: int,
        ir_generator: InterestRateGenerator,
        fx_generator: FXGenerator
    ) -> None:
        self.strike = strike
        self.start = start_val
        self.vol = vol
        self.points_per_path = points_per_path
        self.ir_generator = ir_generator
        self.fx_generator = fx_generator

        return

    def simulate_value(self, paths: int) -> np.array:
        value_paths = [self.simulate_value_single_path() for _ in range(paths)]
        simulated_value_path = np.mean(value_paths, axis=0)
        return simulated_value_path

    def simulate_value_single_path(self) -> np.array:
        stock_path = self._generate_stock_path()
        fx_path = self.fx_generator.get_path(self.points_per_path)

        raw_value_path = self._value_on_path(stock_path)
        ccy_value_path = raw_value_path * fx_path

        return ccy_value_path

    def _generate_stock_path(self) -> np.array:
        """
        Generate a GBM path with stochastic IR, as provided by the classes IR generator.
        Recall that under the risk free measure a stock in the Black-Scholes model has drift
        equal to the interest rate of a risk free asset.
        """
        bm_path = np.random.normal(0.0, self.vol, self.points_per_path)
        ir_path = self.ir_generator.get_path(self.points_per_path)
        
        bm_path += ir_path

        stock_path = np.exp(bm_path)
        stock_path[0] *= self.start
        stock_path = stock_path.cumprod()

        return stock_path
    
    def _value_on_path(self, stock_path: np.array) -> np.array:
        call_value = np.maximum(stock_path - self.strike, np.zeros(self.points_per_path))
        return call_value
    

    

File: rust/equity_call_pricer/src/equity_call_pricer/equity_call_pricer.rs
use crate::{
    random_generator::random_generator::RandomGenerator,
    ir_generator::ir_generator::IRGenerator,
    fx_generator::fx_generator::FXGenerator
};


pub struct EquityCallPricer {
    pub strike: f64,
    pub start: f64,
    pub vol: f64,
    pub points: usize,
    pub rng: RandomGenerator,
    pub ir_generator: IRGenerator,
    pub fx_generator: FXGenerator
}

impl EquityCallPricer {

    pub fn simulate_value_path(&mut self, paths: usize) -> Vec<f64> {
        let mut simulated_value_path: Vec<f64> = vec![0.0; self.points];

        for i in 0..paths {
            let value_path = self.simulate_value_single_path();

            for v in 0..self.points {
                simulated_value_path[v] += (value_path[v] - simulated_value_path[v] ) / (i + 2) as f64;
            }
        }

        simulated_value_path
    }

    pub fn simulate_value_single_path(&mut self) -> Vec<f64> {
        let stock_path = self.generate_stock_path();
        let fx_path = self.fx_generator.get_path(self.points, &mut self.rng);

        let value_path = self.value_on_path(&stock_path);

        value_path.iter().zip(fx_path.iter()).map(|(v, f)| v * f).collect()
    }

    fn generate_stock_path(&mut self) -> Vec<f64> {
        let mut stock_path: Vec<f64> = self.rng.normal(self.points, 0.0, self.vol);
        let ir_path = self.ir_generator.get_path(self.points, &mut self.rng);

        stock_path = stock_path
            .iter()
            .zip(ir_path.iter())
            .map(|(s, r)| f64::exp(*s + r))
            .collect::<Vec<f64>>();

        stock_path
            .iter_mut()
            .fold(self.start, |acc, v| {
            *v *= acc;
            *v
        });

        stock_path
    }

    fn value_on_path(&self, stock_path: &Vec<f64>) -> Vec<f64> {
        stock_path.iter().map(|v| if *v > self.strike {*v} else {0.0}).collect()
    }
}